x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -149,13 +149,19 @@ public void acquire(final EvictionListener listener) throws IOException {
                 synchronized (listeners) {
                     ensureOpen();
                     final boolean added = listeners.add(listener);
-                    assert added : "listener already exists " + listener;
-                    if (listeners.size() == 1) {
-                        assert channelRef == null;
-                        channelRef = new FileChannelReference();
+                    try {
+                        assert added : "listener already exists " + listener;
+                        if (listeners.size() == 1) {
+                            assert channelRef == null;
+                            channelRef = new FileChannelReference();
+                        }
+                        success = true;
+                    } finally {
+                        if (success == false && added) {
+                            listeners.remove(listener);
+                        }
```

Comment:
We synchronize on listeners whenever we access it, so we might as well just add to the listeners list after opening the file? Seems it could avoid this extra try-finally block.

Reply:
Sounds good :) I pushed https://github.com/elastic/elasticsearch/pull/65342/commits/b0570f874ebc49e6f670100a905a00c72e629f15